src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs

```diff
@@ -0,0 +1,176 @@
+                Debug.Assert(_cachingContext != null);
+            }
+
+            return _cachingContext.GetOrAddJsonTypeInfo(type);
+        }
+
+        internal bool TryGetJsonTypeInfo(Type type, [NotNullWhen(true)] out JsonTypeInfo? typeInfo)
+        {
+            if (_cachingContext == null)
+            {
+                typeInfo = null;
+                return false;
+            }
+
+            return _cachingContext.TryGetJsonTypeInfo(type, out typeInfo);
+        }
+
+        internal bool IsJsonTypeInfoCached(Type type) => _cachingContext?.IsJsonTypeInfoCached(type) == true;
+
+        /// <summary>
+        /// Return the TypeInfo for root API calls.
+        /// This has an LRU cache that is intended only for public API calls that specify the root type.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal JsonTypeInfo GetOrAddJsonTypeInfoForRootType(Type type)
+        {
+            JsonTypeInfo? jsonTypeInfo = _lastTypeInfo;
+
+            if (jsonTypeInfo?.Type != type)
+            {
+                jsonTypeInfo = GetOrAddJsonTypeInfo(type);
+                _lastTypeInfo = jsonTypeInfo;
+            }
+
+            return jsonTypeInfo;
+        }
+
+        internal void ClearCaches()
+        {
+            _cachingContext?.Clear();
+            _lastTypeInfo = null;
+        }
+
+        private void InitializeCachingContext()
+        {
+#if NETCOREAPP
+            // Searches all tracked instances for an initialized instance using
+            // equivalent configuration. If found, reuse the caching context here.
+            foreach (KeyValuePair<JsonSerializerOptions, object?> entry in TrackedOptionsInstances.All)
+            {
+                JsonSerializerOptions otherOptions = entry.Key;
+                if (otherOptions._cachingContext != null && EqualityComparer.AreEquivalent(this, otherOptions))
+                {
+                    _cachingContext = otherOptions._cachingContext;
+                    return;
+                }
+            }
+#endif
+            _cachingContext = new(this);
+        }
+
+        /// <summary>
+        /// Stores and manages all reflection caches for one or more <see cref="JsonSerializerOptions"/> instances.
+        /// NB the type encapsulates the original options instance and only consults that one when building new types;
+        /// this is to prevent multiple options instances from leaking into the object graphs of converters which
+        /// could break user invariants.
+        /// </summary>
+        private sealed class CachingContext
+        {
+            private readonly ConcurrentDictionary<Type, JsonConverter> _converterCache = new();
+            private readonly ConcurrentDictionary<Type, JsonTypeInfo> _jsonTypeInfoCache = new();
+
+            public CachingContext(JsonSerializerOptions options)
+            {
+                Options = options;
+                _ = Count;
```

Comment:
`Count` is only accessed via reflection in testing. Added this to prevent the member from being removed in IL but maybe there's a better way to do this?

Reply:
`Count` is more expensive than you'd think, as it requires taking one or more locks.  We shouldn't be calling it in product src just to enable rooting the property for testing purposes.  This should probably have a debug-only trimmer xml file that keeps Count rooted only in debug, or if we really want to avoid that, a DynamicDependency from the ctor to Count (but that will still end up keeping the property rooted in production when it's only needed for testing).
cc: @eerhardt

Reply:
The way we do this is with a `ILLink.Descriptors.LibraryBuild.xml` file to preserve this property during our build. Here's our documentation:

https://github.com/dotnet/runtime/blob/main/docs/workflow/trimming/ILLink-files.md#illinkdescriptorsxml

See

https://github.com/dotnet/runtime/blob/a14d27ac7ff3a58e39d889db72f7ce682c7492fd/src/libraries/System.Private.CoreLib/src/ILLink/ILLink.Descriptors.LibraryBuild.xml#L3-L5

as an example.

Reply:
@eerhardt what's the correct syntax for nested types? I tried adding a file via https://github.com/dotnet/runtime/pull/65376/commits/68bcb5384f01bba7e139380853cb6f1c7e44e389 but it looks like it's still trimming the property.

Reply:
I guess it is a `/` 🤷 

See https://github.com/dotnet/linker/blob/main/docs/data-formats.md#preserve-a-type